### DISCUSSION TITLE
Multiple Bluetooth controllers on Pi Zero-W

### DISCUSSION THREAD

**Discussion by maxlyth** [mKeRix/room-assistant#451] [Q&A]
I have a number of Pi Zero-W deployed around the house each with an external USB Bluetooth dongle so that I can use both the BluetoothClassic integration on the built-in BT controller and BluetoothLE integration on the USB BT controller.

At some point over the last few month the built-in BT controllers stopped working on all the Pi Zeros. If the USB controller is plugged in after the Zero has booted then both controllers show. This looks due a change in Raspbian committed this year to the /usr/bin/btuart file.

The updated script exits if the onboard controller is already initialised but does this by incorrectly assuming that hci0 is the onboard BT controller which is not true if booted with a USB BT device. This can be fixed by only exiting if the MAC address of the BT controller matches a known address for the Pi Foundation controllers. 

So in line 14ish of /usr/bin/btuart change:
```
if [ -e /sys/class/bluetooth/hci0 ]; then
```
to:
```
if ( /usr/bin/hcitool dev | grep -q -E '\s(B8:27:EB:|DC:A6:32:)' ); then
```

This took me a while to figure out so I hope it helps someone else.

**Comment by mKeRix** [comment]
I came across this issue this week with my test setup as well! The relevant script already received a fix in the relevant git repository, but unfortunately they have not released it as a package update yet. An alternative to updating line manually is pulling the fixed script directly from the repository:

```
sudo wget -O /usr/bin/btuart https://raw.githubusercontent.com/RPi-Distro/pi-bluetooth/ede34901ec30342137d2bcde0c2124222d5d1229/usr/bin/btuart
```

Then reboot the Pi. Once they release an updated version this file will just be replaced with the fixed version (plus any other changes that they might do until then), so there is nothing else to consider.

Either way, thanks for posting this detailed write up! I've pinned it for more visibility.

**Comment by toreupfeet** [comment]
This is very useful info. Thank you. Does the external USB dongle provide wider coverage than the built-in BT?

**Reply by mKeRix** [reply]
Depends on the dongle - but I think most provide the same coverage as the internal chip. It does give you the advantage of having a dedicated Bluetooth chip (internal one is shared between WiFi and BT) and more freedom for positioning though. :)

**Reply by k41zen** [reply]
Can someone post the models of external USB Bt dongles that they're using please? I'm finding the internal one and the WIFI are resulting in hci0 disappearing and reappearing every few seconds.
